File: Data Structures & Algorithms/search-for-word/submission-4.java
class Solution {
    public boolean exist(char[][] board, String word) {

        int ROW = board.length;
        int COL = board[0].length;

        Set<String> visited = new HashSet<String>();


        for(int i=0;i<ROW;i++)  {
            for(int j=0;j<COL;j++)  {
                if(dfs(board, word, i, j, 0, visited)) 
                    return true;
        }
        }

        return false;
        
    }

    public boolean dfs(char[][] board, String word, int r, int c, int i, Set<String> set) {
        // true condition
        if(i == word.length()) 
            return true;

        // false condition 
        if(r < 0 || c < 0 || r >= board.length || c >= board[0].length || board[r][c] != word.charAt(i) || set.contains(r + "." + c)) {
            return false;
        }
        // do the visited set here 

        set.add(r + "." + c);

        // perform the dfs 

        boolean res = dfs(board, word, r+1, c, i+1, set) || dfs(board, word, r-1, c, i+1, set)  || dfs(board, word, r, c+1, i+1, set)  || dfs(board, word, r, c-1, i+1, set) ;
        
        // remove visited path 
        set.remove(r + "." + c);
        // return results
        return res;
    }
}

File: Data Structures & Algorithms/longest-palindromic-subsequence/submission-4.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        len_string = len(s)

        res = [[0]*len_string for _ in range(len_string)]

        ROW = len(res)
        COL = len(res[0])

        for i in range(ROW):
            res[i][i] = 1
            
        for r in range(len_string - 1, -1, -1):
            for c in range(r + 1, len_string):
                if s[r] == s[c]:
                    res[r][c] = 2 + res[r+1][c-1]
                else:
                    res[r][c] = max(res[r][c-1], res[r+1][c])

    

        return res[0][len_string-1]


        #T O T A L
        